man/setquota.8
.TH SETQUOTA 8 "23 Jan 1989" "Project Athena"
\" RCSID: $HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/man/setquota.8 $ $Id: setquota.8 3956 2010-01-05 20:56:56Z zacheiss $
.UC 4
.SH NAME
setquota \- set the disk usage quotas for a given user and partition
.SH SYNOPSIS
.B setquota
[-u]
.I special uid quota
.br
.B setquota
-r [-u]
.I special low_uid high_uid quota
.SH DESCRIPTION
.I Setquota
will set the hard and soft limits on disk usage and inode
usage for the user with
.I uid
on the partition
.I special. special
can be either the block device name (ie.
.I /dev/rz3g
) or the directory it's mounted on. (ie.
.I /u1
). The user's soft disk usage limit will be set to
.I quota.
The soft inode usage limit will be set to half that amount.
The hard limits will be set to 1.2 times the soft limits.
.SH OPTIONS
If the
.B \-u
option is given, then the specified
.I quota
will be added to the current usage to obtain the new soft limit.  The
remaining values are then set relative to this as specified above.
If the
.B \-r
option is given, then all the quotas for the inclusive range of uid's
are set to the given quota (plus usage with the
.B \-u
option).
.SH DIAGNOSTICS
.B setquota
will return a non-zero status if the user is not root, if the uid is
not between 0 and 32767 inclusive, if the low-uid is higher than the
high-uid, if the quota is less than zero, or if quotas are not enabled
on the given device or partition.

The previous bug which required the user to manually grow the quotas
file is fixed.
.B setquota
automatically grows the quotas file if necessary.
.SH AUTHORS
Bill Sommerfeld, MIT Project Athena
.br
Jim Rees, MIT Project Athena
.SH "SEE ALSO"
quota(1), quota(2), edquota(8), quotacheck(8), quotaon(8), repquota(8)


debian/changelog
debathena-moira (4.0.0+svn20100225-0debathena2) UNRELEASED; urgency=low

  * There is no "lib" section - it's "libs".

 -- Evan Broder <broder@mit.edu>  Fri, 26 Feb 2010 12:50:19 -0500

debathena-moira (4.0.0+svn20100225-0debathena1) unstable; urgency=low

  * New SVN snapshot.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 15:54:26 -0500

debathena-moira (4.0.0+svn20100223-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Drop patch install-headers, which was incorporated upstream.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 02:13:44 -0500

debathena-moira (4.0.0+svn20100104-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Build and install libmrclient as a shared library.
  * Deal with rename of /etc{/athena => }/moira.conf
  * Patch manpages to deal with our installation paths.

 -- Evan Broder <broder@mit.edu>  Mon, 04 Jan 2010 22:05:10 -0500

debathena-moira (4.0.0+cvs20091116-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop patch fix-update-ticket, which was incorporated upstream.
  * Add a missing definition of ${prefix} in include/Makefile.in so that
    the moira.h header actually gets installed in
    /usr/include/moira. (Trac: #445)

 -- Evan Broder <broder@mit.edu>  Thu, 19 Nov 2009 10:53:40 -0500

debathena-moira (4.0.0+cvs20090922-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Add patch to fix builds without krb4.

 -- Evan Broder <broder@mit.edu>  Tue, 22 Sep 2009 00:35:12 -0400

debathena-moira (4.0.0+cvs20090728-0debathena3) unstable; urgency=low

  * Add a missing conflicts to make upgrades run more smoothly.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 18:28:33 -0700

debathena-moira (4.0.0+cvs20090728-0debathena2) unstable; urgency=low

  * Include the headers in debathena-libmoira-dev.
  * Install update_test in moira-clients instead of moira-update-server.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 17:54:34 -0700

debathena-moira (4.0.0+cvs20090728-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Fix zephyr dependencies to actually pull in Kerberized libraries; drop
    build-dep on Kerberized library.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 13:43:28 -0400

debathena-moira (4.0.0+cvs20090621-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop krb4less-fix; it was incorporated upstream.
  * Universally stop building Moira with krb4 support.

 -- Evan Broder <broder@mit.edu>  Fri, 26 Jun 2009 00:51:28 -0400

debathena-moira (4.0.0+cvs20090618-0debathena1) unstable; urgency=low

  * New upstream release
  * Build and install libmoira as a shared library. (Trac: #70)
  * Drop most of the krb4 patch - it's been incorporated upstream.
  * Install the Moira development headers by patching the relevant
    Makefiles, instead of in the debian/rules file.

 -- Evan Broder <broder@mit.edu>  Sat, 20 Jun 2009 14:45:32 -0400

debathena-moira (4.0.0+cvs20090409-0debathena1) unstable; urgency=low

  * New CVS snapshot (Trac: #195)
  * Drop patches that have been incorporated upstream.
  * Add patch to allow building on systems without krb4. (Trac: #198)

 -- Evan Broder <broder@mit.edu>  Sun, 12 Apr 2009 14:03:22 -0400

debathena-moira (4.0.0+cvs20080830-0debathena2) unstable; urgency=low

  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 15:32:35 -0500

debathena-moira (4.0.0+cvs20080830-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Add a debathena-moira-update-server package that installs
    update_server

 -- Evan Broder <broder@mit.edu>  Fri, 01 Aug 2008 10:54:26 -0400

debathena-moira (4.0.0+cvs20080701-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Move Moira header files from /usr/share/include to /usr/include/moira

 -- Evan Broder <broder@mit.edu>  Tue, 01 Jul 2008 20:38:13 -0400

debathena-moira (4.0.0+cvs20080304-0debathena1) unstable; urgency=low

  * New CVS snapshot to fix chsh.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 04 Mar 2008 15:50:31 -0500

debathena-moira (4.0.0+cvs20080301-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Remove mrtest segfault patch from last release (fixed in upstream)

 -- Tim Abbott <tabbott@mit.edu>  Sat,  1 Mar 2008 18:51:40 -0500

debathena-moira (4.0.0+cvs20070817-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Patch an mrtest segfault with too many command line arguments.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 24 Aug 2007 19:53:49 -0400

debathena-moira (4.0.0+cvs20070129-0debathena3) unstable; urgency=low

  * Move chsh.debathena to chsh.moira and chfn.debathena to chfn.moira.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 06 Aug 2007 03:17:10 -0400

debathena-moira (4.0.0+cvs20070129-0debathena2) unstable; urgency=low

  * Add missing build dependencies to libmoira-dev.
  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:34:07 -0400

debathena-moira (4.0.0+cvs20070129-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 14 Feb 2007 00:53:24 -0500


gen/setquota.c
/*      $Id: setquota.c 3956 2010-01-05 20:56:56Z zacheiss $
 *
 *      Copyright (C) 1987 by the Massachusetts Institute of Technology
 *
 *      Set quota on specified device for specified user to specified value.
 *
 *      Uses the NFS style quota system/quotactl rather than the Melbourne
 * quota system.
 *
 */

#ifndef lint
static char *rcsid_setquota_c = "$HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/gen/setquota.c $ $Id: setquota.c 3956 2010-01-05 20:56:56Z zacheiss $";
#endif lint

#include <stdio.h>
#include <ctype.h>
#include <mntent.h>

#include <sys/file.h>
#include <sys/param.h>
#include <sys/time.h>
#include <ufs/quota.h>

#define kb(n)	(howmany(dbtob(n), 1024))

static char device[20];
static char quotafilename[30];
static struct dqblk zblk = {0};

int main(int argc, char **argv)
{
  int uid, uid_low, uid_high, soft_quota, qfd;
  struct dqblk db, odb;
  int uflag = 0;
  int range_mode = 0;

  while (argc > 4 && *argv[1] == '-')
    {
      switch (argv[1][1])
	{
	case 'u':
	  uflag = 1;
	  --argc;
	  ++argv;
	  break;
	case 'r':
	  range_mode = 1;
	  --argc;
	  ++argv;
	  break;
	default:
	  goto usage;
	}
    }

  if ((argc != 4 && !range_mode) || (argc != 5 && range_mode))
    {
    usage:
      fprintf(stderr, "usage: setquota [-u] special uid quota\n"
	      "setquota -r [-u] special uid_low uid_high quota\n"
	      "-u means set limit to <quota> + cur usage\n"
	      "special is a mounted filesystem special device\n"
	      "quota is in 1KB units\n");
      exit(1);
    }

  if ((!range_mode && (!isdigit(*argv[2]) || !isdigit(*argv[3]))) ||
      (range_mode && (!isdigit(*argv[2]) || !isdigit(*argv[3]) ||
		      !isdigit(*argv[4]))))
    {
      fprintf(stderr, "setquota: uid and quota must be numeric\n");
      goto usage;
    }

  if (range_mode)
    {
      uid_low = atoi(argv[2]);
      uid_high = atoi(argv[3]);
      soft_quota = atoi(argv[4]);
      if (uid_low > uid_high)
	{
	  fprintf(stderr, "setquota: range error\n");
	  exit(1);
	}
    }
  else
    {
      uid_low = uid_high = atoi(argv[2]);
      soft_quota = atoi(argv[3]);
    }

  get_device(argv[1]);

  for (uid = uid_low; uid <= uid_high; uid++)
    {
      if (quotactl(Q_GETQUOTA, device, uid, &odb))
	{
	  if (!(qfd = open(quotafilename, O_RDWR)))
	    {
	      perror("No quota file");
	      exit(1);
	    }

	  lseek(qfd, 32767 * sizeof(struct dqblk), L_SET);
	  write(qfd, &zblk, sizeof(struct dqblk));
	  close(qfd);

	  if (quotactl(Q_GETQUOTA, device, uid, &odb) != 0)
	    {
	      perror("Can't get current quota info");
	      exit(1);
	    }
	}

      db.dqb_bsoftlimit = soft_quota;
      db.dqb_bhardlimit = (db.dqb_bsoftlimit * 6) / 5;
      db.dqb_fsoftlimit = soft_quota / 2;
      db.dqb_fhardlimit = (db.dqb_fsoftlimit * 6) / 5;
      db.dqb_btimelimit = odb.dqb_btimelimit;
      db.dqb_ftimelimit = odb.dqb_ftimelimit;

      db.dqb_bsoftlimit *= btodb(1024);
      db.dqb_bhardlimit *= btodb(1024);

      if (uflag)
	{
	  db.dqb_bhardlimit += odb.dqb_curblocks;
	  db.dqb_bsoftlimit += odb.dqb_curblocks;
	  db.dqb_fhardlimit += odb.dqb_curfiles;
	  db.dqb_fsoftlimit += odb.dqb_curfiles;
	}

      if (quotactl(Q_SETQLIM, device, uid, &db) < 0)
	{
	  fprintf(stderr, "quotactl: %d on ", uid);
	  perror(device);
	  exit(1);
	}
    }

  if (quotactl(Q_SYNC, device, 0, 0) < 0)
    {
      perror("can't sync disk quota");
      exit(1);
    }

  exit(0);
}

get_device(char *device_or_dir)
{
  struct mntent *mntp;
  FILE *fstab;

  fstab = setmntent(MNTTAB, "r");
  while (mntp = getmntent(fstab))
    {
      if (!strcmp(mntp->mnt_fsname, device_or_dir) ||
	  !strcmp(mntp->mnt_dir, device_or_dir))
	{
	  strcpy(device, mntp->mnt_fsname);
	  sprintf(quotafilename, "%s/quotas", mntp->mnt_dir);
	  endmntent(fstab);
	  return;
	}
    }
  fprintf(stderr, "%s not mounted.\n", device_or_dir);
  exit(1);
}
